src/module_3/assign_3.md

```diff
@@ -0,0 +1,980 @@
+    '''
+    fig, ax = plt.subplots(1, 2, figsize=(10, 4))
+    fpr, tpr, thresholds = metrics.roc_curve(real_y, y_prob)
+    roc_auc = metrics.auc(fpr, tpr)
+    
+    fig.suptitle(f'ROC and Precision-Recall Curves: {title}')
+
+    ax[0].set_title(f'ROC curve (AUC = {roc_auc:.2f})')
+    ax[0].plot(fpr, tpr, color='darkorange', lw=2)
+    ax[0].plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')
+    ax[0].set_xlim([0.0, 1.0])
+    ax[0].set_ylim([0.0, 1.05])
+    ax[0].set_xlabel('False Positive Rate')
+    ax[0].set_ylabel('True Positive Rate')
+
+    precision, recall, thresholds = metrics.precision_recall_curve(real_y, y_prob)
+
+    average_precision = metrics.average_precision_score(real_y, y_prob)
+
+    ax[1].plot(recall, precision, marker='.')
+    ax[1].set_xlabel('Recall')
+    ax[1].set_ylabel('Precision')
+    ax[1].set_title('Precision-Recall Curve (AUC-PR = {:.2f})'.format(average_precision))
+    plt.tight_layout()
+    plt.show()
+
+
+```
+
+
+```python
+#Base model with no penalties
+log_model_base = LogisticRegression(penalty= None, n_jobs=-1, class_weight='balanced').fit(train_data_x,train_data_y)
+
+y_train_pred_base = log_model_base.predict(train_data_x)
+y_train_prob_base = log_model_base.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_base,"No regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_base))
+
+```
+
+
+    
+![png](assign_3_files/assign_3_12_0.png)
+    
+
+
+    [[1394040       0]
+     [  21434       0]]
+
+
+
+```python
+def best_threshold(y_true:pd.core.series.Series, y_prob:np.ndarray)->int:
+    '''
+    Function to obtain the threshold that minimizes false positive to be the closest possible to 1%.
+    '''
+    mean = y_prob.mean()
+    std = y_prob.std()
+    thresholds = np.arange(mean - 3*std, mean + 3*std, std/5)
+
+    best_threshold = None
+
+    for threshold in thresholds:
+
+        y_pred = (y_prob >= threshold).astype(int)
+        
+        false_positives = metrics.confusion_matrix(y_true, y_pred)[0][1]
+
+        if false_positives > y_true.sum()/100:
+            best_threshold = threshold
+
+    print(f"Best Threshold: {best_threshold}")
+
+    return best_threshold
+```
+
+
+```python
+threshold_base = best_threshold(train_data_y, y_train_prob_base)
+```
+
+    Best Threshold: 0.49994339123985077
+
+
+
+```python
+#Model with l2 Regularization
+
+param_grid = {
+    'C': [10 ** i for i in range(-4,3)],
+    'penalty': ['l2']}
+
+log_model_reg = LogisticRegression(n_jobs=-1, class_weight='balanced')
+
+grid_search_l2 = GridSearchCV(log_model_reg, param_grid, cv=5, scoring='accuracy',n_jobs=-1).fit(train_data_x,train_data_y)
+
+print(grid_search_l2.best_params_)
+
+y_train_pred_l2 = grid_search_l2.predict(train_data_x)
+y_train_prob_l2 = grid_search_l2.predict_proba(train_data_x)[:,1]
+
+draw_roc_prc(train_data_y, y_train_prob_l2,"With regularization")
+print(metrics.confusion_matrix(train_data_y,y_train_pred_l2))
```

Comment:
Your model still not learning anything. Would look at `model.coef_`. It's likely the case that the solver isn't converging because you are not scaling/standardising the data. There may be some warning/error being printed potentially.

Reply:
I didn't get any warnings, but the coefficients were all getting the same values close to 0.